serverless/internal/storage/fs/fs.go

```diff
@@ -102,21 +116,123 @@ func (fs *Storage) LogState() api.LogState {
 func (fs *Storage) Sequence(leafhash []byte, leaf []byte) error {
-	return errors.New("unimplemented")
+	// 1. Check for dupe leafhash
+	// 2. Write temp file
+	// 3. Hard link temp -> seq file
+	// 4. Optimistically symlink leafhash -> seq file
+
+	// Ensure the leafhash diretory structure is pressent
+	leafDir, leafFile := layout.LeafPath(fs.rootDir, leafhash)
+	if err := os.MkdirAll(leafDir, dirPerm); err != nil {
+		return fmt.Errorf("failed to make leaf directory structure: %w", err)
+	}
+	// Check for dupe leaf already present.
+	leafFQ := filepath.Join(leafDir, leafFile)
+	if _, err := os.Stat(leafFQ); os.IsExist(err) {
+		glog.Infof("Skipping dupe leafhash 0x%x", leafhash)
+		return nil
+	}
+
+	tmp := filepath.Join(fs.rootDir, fmt.Sprintf(leavesPendingPathFmt, leafhash))
+	if err := createExcl(tmp, leaf); err != nil {
+		return fmt.Errorf("unable to write temporary file: %w", err)
+	}
+	defer func() {
+		os.Remove(tmp)
+	}()
+
+	// Now try to sequence it, we may have to scan over some newly sequenced entries
+	// if Sequence has been called since the last time an Integrate/UpdateState
+	// was called.
+	for {
+		seq := fs.nextSeq
+
+		// Ensure the sequencing directory structure is present:
+		seqDir, seqFile := layout.SeqPath(fs.rootDir, seq)
+		if err := os.MkdirAll(seqDir, dirPerm); err != nil {
+			return fmt.Errorf("failed to make seq directory structure: %w", err)
+		}
+
+		// Hardlink the sequence file to the temporary file
+		seqPath := filepath.Join(seqDir, seqFile)
+		if err := os.Link(tmp, seqPath); errors.Is(err, os.ErrExist) {
+			// That sequence number is in use, try the next one
+			fs.nextSeq++
+			continue
+		} else if err != nil {
+			return fmt.Errorf("failed to link seq file: %w", err)
+		}
+
+		// Finally, try to link leafhash -> seq.
+		err := os.Symlink(seqPath, leafFQ)
+		if err != nil && !errors.Is(err, os.ErrExist) {
+			return fmt.Errorf("failed to link leafdata file: %w", err)
+		}
+
+		// All done!
+		break
+	}
+
+	return nil
+}
+
+// createExcl, creates the named file before writing the data in d to it.
+// It will error if the file already exists, or it's unable to fully write the
+// data & close the file.
+func createExcl(f string, d []byte) error {
+	tmpFile, err := os.OpenFile(f, os.O_RDWR|os.O_CREATE|os.O_EXCL, filePerm)
+	if err != nil {
+		return fmt.Errorf("unable to create temporary file: %w", err)
+	}
+	n, err := tmpFile.Write(d)
+	if err != nil {
+		return fmt.Errorf("unable to write leafdata to temporary file: %w", err)
+	}
+	if got, want := n, len(d); got != want {
+		return fmt.Errorf("short write on leaf, wrote %d expected %d", got, want)
+	}
+	if err := tmpFile.Close(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// ScanSequenced calls the provided function once for each contiguous entry
+// in storage starting at begin.
+// The scan will abort if the function returns an error.
+func (fs *Storage) ScanSequenced(begin uint64, f func(seq uint64, entry []byte) error) (uint64, error) {
```

Comment:
Document the return value. It could be the last index or the number of sequenced entries. (it's the number of sequenced entries, but I only know from reading the impl)

Reply:
done